querqy-core/src/main/java/querqy/model/rewriting/RewriterOutput.java

```diff
@@ -0,0 +1,51 @@
+package querqy.model.rewriting;
+
+import querqy.model.ExpandedQuery;
+import querqy.model.logging.ActionLogging;
+import querqy.model.logging.RewriterLogging;
+
+import java.util.List;
+import java.util.Optional;
+
+public class RewriterOutput {
+
+    private final ExpandedQuery expandedQuery;
+    private final RewriterLogging rewriterLogging;
+
+    private RewriterOutput(final ExpandedQuery expandedQuery, final RewriterLogging rewriterLogging) {
+        this.expandedQuery = expandedQuery;
+        this.rewriterLogging = rewriterLogging;
+    }
+
+    public ExpandedQuery getExpandedQuery() {
+        return expandedQuery;
+    }
+
+    public Optional<RewriterLogging> getRewriterLogging() {
+        return Optional.ofNullable(rewriterLogging);
+    }
+
+    public static RewriterOutputBuilder builder() {
+        return new RewriterOutputBuilder();
+    }
+
+    public static class RewriterOutputBuilder {
```

Comment:
As long as we use a builder we can enhance the output without any refactoring regarding existing tests (I want to avoid the need to refactor tons of tests just to let them additionally pass null). If we have multiple constructors, it is not obvious, which one is valid for regular use and which ones should only be used for tests. With a builder, this is very simple: You need to pass everything as long they are not defined as Optional<> in the underlying context object. For pure testing purposes, you can leave them out. 

This is why I would avoid using constructors whenever possible.

Reply:
Still the builder makes you more flexible in refactoring the structure of the context object (which is definitely going to happen for decorations in the future). I think there are valid arguments on both sides, so the decision is somehow a matter of personal preference.